### ISSUE TITLE
review development environment requirements

### ISSUE THREAD

**Issue by daverodgman** [ARMmbed/mbedtls#4379]
Look at minimum versions of tools required to build Mbed TLS and ensure that we (a) aren't limiting ourselves by requiring very old versions of tools that are missing useful features, and (b) if we make changes, that the new minimum version is not so new as to exclude common development environments.

cmake is the obvious candidate for a minimum version bump (to get package export features as per #4259) but we should also consider Perl, Python, sh, Make, Visual Studio, clang, gcc, etc.

If (b) is limiting us, we could consider making changes on 3.0 only.

[This should be documented](https://github.com/ARMmbed/mbedtls/issues/2896) (there are some docs in README.md already - possibly more details could be added as it seems to be missing the cmake version, and doesn't discuss platforms - e.g. which old versions of Ubuntu we want to support).

**Comment by gilles-peskine-arm** [comment]
See https://github.com/ARMmbed/mbedtls/issues/2896 for last year's discussion, and in particular https://github.com/ARMmbed/mbedtls/issues/2896#issuecomment-603471273 for a list of tool versions on some popular long-term-support operating systems.

There's a [2021 mailing list post on the topic](https://lists.trustedfirmware.org/pipermail/mbed-tls/2021-March/000319.html), which after 3 weeks has had no reply.

**Comment by gilles-peskine-arm** [comment]
CMake is the only place where we'd really like to move forward. For other tools:

* Perl: version irrelevant. We only use perl for old scripts that run under very old versions.
* Python: we've already required 3.4 for a while. Newer versions (3.5, 3.8) incrementally have better support for types, but that's only a minor gain.
* sh: version irrelevant. We only use portable sh.
* Make: version irrelevant as long as it's GNU make.
* Visual Studio: We need at least VS 2012 for C99, and we've never tested with VS 2012 so we already require VS 2013 or above. Last I checked, I didn't see anything to gain by moving to newer versions, except maybe VS 2019's C99 support has finally improved but it's too early to drop VS 2017 anyway.
* clang, gcc: versions only relevant for the default flags in Makefile. In terms of C code, we merely require a C99 compiler.
